proxy/proxy.ts

```diff
@@ -13,17 +13,179 @@ class BadRequestError extends Error {
   }
 }
 
-export function createProxyServer(config: {
+interface RateLimitInfo {
+  count: number
+  timestamp: number
+}
+
+interface ProxyConfig {
   allowLocalhost?: boolean
   allowsFrom: RegExp[]
   maxSize: number
   timeout: number
-}): Server {
-  return createServer(async (req, res) => {
+}
+
+const RATE_LIMIT = {
+  PER_DOMAIN: {
+    LIMIT: 500,
+    DURATION: 60 * 1000
+  },
+  GLOBAL: {
+    LIMIT: 5000,
+    DURATION: 60 * 1000
+  }
+}
+
+const delay = (ms: number): Promise<void> => {
+  return new Promise(resolve => setTimeout(resolve, ms))
+}
+
+const rateLimitMap: Map<string, RateLimitInfo> = new Map()
+const requestQueue: Map<string, Promise<void>> = new Map()
+
+const isRateLimited = (ip: string, domain: string): boolean => {
+  const now = performance.now()
+
+  const domainKey = `${ip}:${domain}`
+  const domainRateLimit = rateLimitMap.get(domainKey) || {
+    count: 0,
+    timestamp: now
+  }
+
+  if (now - domainRateLimit.timestamp > RATE_LIMIT.PER_DOMAIN.DURATION) {
+    domainRateLimit.count = 0
+    domainRateLimit.timestamp = now
+  }
+
+  if (domainRateLimit.count >= RATE_LIMIT.PER_DOMAIN.LIMIT) {
+    return true
+  }
+
+  const globalKey = ip
+  const globalRateLimit = rateLimitMap.get(globalKey) || {
+    count: 0,
+    timestamp: now
+  }
+
+  if (now - globalRateLimit.timestamp > RATE_LIMIT.GLOBAL.DURATION) {
+    globalRateLimit.count = 0
+    globalRateLimit.timestamp = now
+  }
+
+  if (globalRateLimit.count >= RATE_LIMIT.GLOBAL.LIMIT) {
+    return true
+  }
+
+  domainRateLimit.count += 1
+  globalRateLimit.count += 1
+  rateLimitMap.set(domainKey, domainRateLimit)
+  rateLimitMap.set(globalKey, globalRateLimit)
+
+  return false
+}
+
+const handleError = (e: any, res: ServerResponse): void => {
```

Comment:
```suggestion
const handleError = (e: unknown, res: ServerResponse): void => {
```

Reply:
I guess [that](https://github.com/hplush/slowreader/pull/227/commits/729ee25fa33adede8dea6bcdeb110172d4bb6973) should do the job